src/engine.ml

```diff
@@ -1918,6 +1892,101 @@ let incoming state control_crypto buf =
+        | Error _, Ok crypt -> Ok (`Tls_crypt (crypt, None))
+        | _ -> Error error
+      in
+      ( {
+          config = server.server_config;
+          is_not_taken = server.is_not_taken;
+          control_crypto :> control_crypto;
+          state = Server Server_handshaking;
+          linger = Cstruct.empty;
+          rng = server.server_rng;
+          ts = server.server_ts;
+          now = server.server_now;
+          session;
+          channel;
+          lame_duck = None;
+          last_received = current_ts;
+          last_sent = current_ts;
+        },
+        control_crypto,
+        data )
+    in
+    match Config.find Tls_crypt_v2_server server.server_config with
+    | None ->
+        tls_auth_or_tls_crypt
+          (`Msg "server only supports tls-auth or tls-crypt")
+    | Some (_wrapping_key, true) ->
+        (* TODO: HMAC cookie support *)
+        Error (`Msg "Server does not support hmac cookies (yet)")
+    | Some (wrapping_key, false) -> (
+        match Packet.decode_key_op protocol data with
+        | Error `Tcp_partial ->
+            (* It is unlikely that we don't read a full packet in first try and it is annoying to handle. *)
+            assert false
+        | Error _ as e -> e
+        | Ok (Packet.Hard_reset_client_v2, _key, _received, linger) ->
+            assert (Cstruct.is_empty linger);
+            tls_auth_or_tls_crypt (`Msg "server only supports tls-crypt-v2")
+        | Ok (Packet.Hard_reset_client_v3, _key, received, linger) ->
+            assert (Cstruct.is_empty linger);
+            (* decode and unwrap wKc *)
+            let* actual_packet, wkc =
+              Tls_crypt.Wrapped_key.of_cstruct received
+            in
+            let* tls_crypt, _metadata =
+              Tls_crypt.Wrapped_key.unwrap ~key:wrapping_key wkc
+            in
+            let control_crypto =
+              `Tls_crypt
+                ( {
+                    my = Tls_crypt.server_key tls_crypt;
+                    their = Tls_crypt.client_key tls_crypt;
+                  },
+                  None )
+            in
+            Ok
+              ( {
+                  config = server.server_config;
+                  is_not_taken = server.is_not_taken;
+                  control_crypto;
+                  state = Server Server_handshaking;
+                  linger = Cstruct.empty;
+                  rng = server.server_rng;
+                  ts = server.server_ts;
+                  now = server.server_now;
+                  session;
+                  channel;
+                  lame_duck = None;
+                  last_received = current_ts;
+                  last_sent = current_ts;
+                },
+                control_crypto,
+                actual_packet )
+        | Ok (_not_hard_reset_client, _key, _received, _linger) ->
+            Error (`Msg "invalid initial packet"))
+    (* `No_transition *)
+  in
+  incoming t control_crypto packet
```

Comment:
in incoming, there's:
```OCaml
              | Packet.Hard_reset_client_v3 ->
                  ignore_udp_error (Error (`No_transition (ch, op, received)))
```

which we should remove now, shouldn't we? and the serve should then also in `incoming_control_server` deal with reset_3...